blog/src/a-base32-checksum.md

```diff
@@ -0,0 +1,1241 @@
+In his words:
+
+> It is commonly known that in copying a number comprised of a plurality of
+> digits it often happens that an error occurs by transposing two of the
+> digits.
+
+Luhn's goal was to detect all single-digit errors and most transpositions,
+with a single check digit. However, it missed some transpositions, such as
+09 / 90.
+
+<div id=luhnWidget>
+  <input class=luhnInput value=1234590>
+  <strong><output class=luhnValidity>Valid</output></strong>
+  (Add a <strong><output class=luhnDigit>6</output></strong> to be valid).
+</div>
+
+<script>
+// From https://github.com/EDumdum/luhn/blob/master/src/luhn.js
+function luhnValid(value) {
+  return /^[0-9]+$/.test(value) && luhnRemainder(value) === 0;
+}
+function luhnGen(value) {
+  if (!/^[0-9]+$/.test(value)) { value = ''; }
+  return ((10 - luhnRemainder(value + '0')) % 10);
+}
+function luhnRemainder(value) {
+  var array = [0, 2, 4, 6, 8, 1, 3, 5, 7, 9];
+  var length = value.length, accumulator = 0, bit = 0;
+  while (length-- > 0) {
+    accumulator += (bit ^= 1) ? value.charCodeAt(length) - '0'.charCodeAt(0) : array[value.charCodeAt(length) - '0'.charCodeAt(0)];
+  }
+  return accumulator % 10;
+}
+function initLuhnWidget() {
+  const luhnWidget = document.getElementById('luhnWidget');
+  const luhnInput = luhnWidget.getElementsByClassName('luhnInput')[0];
+  const luhnValidity = luhnWidget.getElementsByClassName('luhnValidity')[0];
+  const luhnDigit = luhnWidget.getElementsByClassName('luhnDigit')[0];
+  const updateWidget = () => {
+    const value = luhnInput.value;
+    luhnValidity.textContent = luhnValid(value) ?  'Valid' : 'Invalid';
+    const digit = luhnGen(value);
+    luhnDigit.textContent = (digit !== digit) ? 'removal of non-digits' : digit;
+  };
+  luhnInput.addEventListener('input', updateWidget);
+  updateWidget();
+}
+addEventListener('DOMContentLoaded', initLuhnWidget);
+</script>
+
+Whether a single check digit could also detect all adjacent transpositions
+was an open question, until 1969, when **[Verhoeff][]** designed another
+algorithm that did just that.
+Furthermore, he gathered the results of a study into a thorough analysis
+of the causes of human errors.
+
+Meanwhile, the International Standards Organization published
+[ISO/IEC 7064][] in 1983. It includes algorithms for single- and
+double-digit checksums of numerical and alphanumerical inputs, with the goal
+to detect all single substitutions and most adjacent transpositions.
+
+They intended it for use in future standards.
+For instance, **[IBAN][ISO 13616]** (the International Bank Account Number)
+needed a checksum to lower the risk of accidentally transfering funds to the
+wrong account number.
+
+Surprisingly, the IBAN system, which supports alphanumeric characters,
+relied on MOD 97-10, which was designed only for numeric inputs.
+
+Our analysis indicates that this was a BAD MOVE™.
+
+A root promise of MOD 97-10 as specified in [ISO/IEC 7064][] is to detect
```

Comment:
Oh ok, it gets clearer here but maybe could have been a bit more obvious higher up 👍